File: software/ios/Tucklet/Models/TransferEstimator.swift
// TransferEstimator.swift
// Swift mirror of `tucklet-core::estimate` + `::link`. Same formula, same
// numbers, so the time the app shows matches what the device computes.
// (Production option: share the Rust core directly via UniFFI to eliminate
// any drift; this native mirror keeps the app self-contained and is unit-test
// equivalent to the Rust tests.)
//
// License: PolyForm Noncommercial 1.0.0

import Foundation

public struct LinkProfile: Sendable, Equatable {
    public let sustainedBps: UInt64
    public let perFileOverheadMs: UInt32
    public init(_ sustainedBps: UInt64, _ perFileOverheadMs: UInt32) {
        self.sustainedBps = sustainedBps; self.perFileOverheadMs = perFileOverheadMs
    }

    // Conservative ("you will actually get this") profiles from docs/TRANSFER_PERFORMANCE.md.
    public static let c5Wireless     = LinkProfile(9_000_000, 40)
    public static let dualC5Wireless = LinkProfile(15_000_000, 40)
    public static let s3Wireless     = LinkProfile(4_000_000, 50)
    public static let wiredUsbHs     = LinkProfile(30_000_000, 8)

    public static func profile(radio: RadioKind, transport: DataTransport) -> LinkProfile {
        switch transport {
        case .wiredUsbHs: return .wiredUsbHs
        case .softAp, .wifiAware:
            switch radio { case .singleC5: return .c5Wireless; case .dualC5: return .dualC5Wireless }
        }
    }
}

public struct TransferEstimate: Sendable, Equatable {
    public let seconds: UInt32
    public let bytesTotal: UInt64
    public let files: UInt32

    /// "12s", "3 min", "1 hr 4 min".
    public var human: String { TransferEstimator.human(seconds: seconds) }
}

public enum TransferEstimator {
    /// time = sum_over_files( size / sustained + per_file_overhead )
    /// Per-file overhead is why 30 small photos differ from one big file of the
    /// same total size — and why naive total/speed lies to users.
    public static func estimate(sizes: [UInt64], link: LinkProfile) -> TransferEstimate {
        let files = UInt32(sizes.count)
        let bytesTotal = sizes.reduce(UInt64(0), &+)
        let bps = UInt128(max(link.sustainedBps, 1))
        var ms: UInt128 = 0
        for s in sizes {
            ms += (UInt128(s) * 1000) / bps
            ms += UInt128(link.perFileOverheadMs)
        }
        let seconds = UInt64((ms + 999) / 1000)
        return TransferEstimate(seconds: UInt32(min(seconds, UInt64(UInt32.max))),
                                bytesTotal: bytesTotal, files: files)
    }

    /// Live ETA from measured throughput, so the displayed number stays honest.
    public static func eta(bytesRemaining: UInt64, measuredBps: UInt64) -> UInt32 {
        guard measuredBps > 0 else { return UInt32.max }
        let s = (bytesRemaining + measuredBps - 1) / measuredBps
        return UInt32(min(s, UInt64(UInt32.max)))
    }

    public static func human(seconds total: UInt32) -> String {
        if total < 60 { return "\(total)s" }
        if total < 3600 {
            let m = total / 60, s = total % 60
            return s == 0 ? "\(m) min" : "\(m) min \(s)s"
        }
        let h = total / 3600, m = (total % 3600) / 60
        return m == 0 ? "\(h) hr" : "\(h) hr \(m) min"
    }
}

// UInt128 is available in Swift 6 stdlib. If targeting an older toolchain,
// replace the UInt128 arithmetic above with Double math (acceptable for an
// estimate) — CONFIRM availability against your Swift version.

File: software/android/app/src/main/java/app/tucklet/core/TransferEstimator.kt
// TransferEstimator.kt
// Pure Kotlin mirror of `tucklet-core` (link profiles + estimator + trickle).
// Same constants and formulas as the Rust, which is the tested source of truth.
//
// License: PolyForm Noncommercial 1.0.0
package app.tucklet.core

import app.tucklet.protocol.DataTransport
import app.tucklet.protocol.RadioKind
import kotlin.math.min

/** Throughput model: sustained bytes/sec + a fixed per-file overhead (ms). */
data class LinkProfile(val sustainedBps: Long, val perFileOverheadMs: Int) {
    companion object {
        // Conservative, measured numbers (docs/TRANSFER_PERFORMANCE.md).
        val C5_WIRELESS = LinkProfile(9_000_000, 40)
        val DUAL_C5_WIRELESS = LinkProfile(15_000_000, 40)
        val WIRED_USB_HS = LinkProfile(30_000_000, 8)

        fun profileFor(radio: RadioKind, transport: DataTransport): LinkProfile =
            when (transport) {
                DataTransport.WIRED_USB_HS -> WIRED_USB_HS
                DataTransport.SOFT_AP, DataTransport.WIFI_AWARE -> when (radio) {
                    RadioKind.SINGLE_C5 -> C5_WIRELESS
                    RadioKind.DUAL_C5 -> DUAL_C5_WIRELESS
                }
            }
    }
}

data class Estimate(val seconds: Int, val bytesTotal: Long, val files: Int) {
    val human: String get() = TransferEstimator.humanSeconds(seconds)
}

object TransferEstimator {
    /**
     * time = sum_over_files( size / sustained + per_file_overhead )
     *
     * The per-file overhead is what makes "30 small photos" differ from one big
     * video of the same total size — exactly why `total / speed` lies.
     */
    fun estimate(sizes: List<Long>, link: LinkProfile): Estimate {
        val bps = link.sustainedBps.coerceAtLeast(1)
        var ms = 0L
        for (s in sizes) {
            ms += (s * 1000L) / bps
            ms += link.perFileOverheadMs.toLong()
        }
        val seconds = (ms + 999) / 1000 // round up
        return Estimate(
            seconds = seconds.coerceAtMost(Int.MAX_VALUE.toLong()).toInt(),
            bytesTotal = sizes.sum(),
            files = sizes.size,
        )
    }

    /** Recompute remaining seconds mid-transfer from measured throughput. */
    fun eta(bytesRemaining: Long, measuredBps: Long): Int {
        if (measuredBps <= 0) return Int.MAX_VALUE
        val s = (bytesRemaining + measuredBps - 1) / measuredBps // ceil
        return s.coerceAtMost(Int.MAX_VALUE.toLong()).toInt()
    }

    fun humanSeconds(total: Int): String = when {
        total < 60 -> "${total}s"
        total < 3600 -> {
            val m = total / 60; val s = total % 60
            if (s == 0) "$m min" else "$m min ${s}s"
        }
        else -> {
            val h = total / 3600; val m = (total % 3600) / 60
            if (m == 0) "$h hr" else "$h hr $m min"
        }
    }
}

/** Trickle: solve slow bandwidth with time, not speed — drip new items whenever
 *  it's cheap, so the big transfer never has to happen. */
object Trickle {
    const val MIN_BATTERY_ON_BATTERY = 30

    data class Conditions(
        val phoneInRange: Boolean,
        val phoneIdle: Boolean,
        val charging: Boolean,
        val batteryPercent: Int,
        val pendingItems: Int,
    )

    data class Decision(val shouldRun: Boolean, val batchSize: Int)

    fun decide(c: Conditions): Decision {
        if (c.pendingItems == 0 || !c.phoneInRange) return Decision(false, 0)
        val ok = c.charging || (c.phoneIdle && c.batteryPercent >= MIN_BATTERY_ON_BATTERY)
        if (!ok) return Decision(false, 0)
        val batch = if (c.charging) 25 else 5
        return Decision(true, min(batch, c.pendingItems))
    }
}
